### ISSUE TITLE
[HOLD for payment 2023-03-13] [$1000] Group member orders keeps on changing randomly when it has more then 1 user with phone number as primaryLogin

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14886]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any group report with more then 1 user with phone number as primaryLogin or create a group with 2 or more users having phone number as primaryLogin
3. Click on profile Icons of members displayed on top of group report chat
4. Click on one of the phone number user and click on message
5. Navigate back to report archive page
6. Observe members order in name of group and open the group to observe the profile icons of members on top of group chat

## Expected Result:
Members order should remain same as it was before

## Actual Result:
Members order changes to phone number first and then email or name and when we open the group chat, it glitches and rectifies the members order as back to normal

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.66-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/217926031-1ffc3fd8-4917-4452-8104-cfca541342e2.mp4

https://user-images.githubusercontent.com/43996225/219061754-03239976-74ab-4dfe-be5d-dd197c610d57.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675703922896249

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0128253986dcc547e6</li>
        <li>Upwork Job ID: 1629147765832331264</li>
        <li>Last Price Increase: 2023-02-24</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kavimuru** [comment]
@slafortune My bad, updated the steps

**Comment by slafortune** [comment]
Thanks, @kavimuru Is this happening on all platforms?

**Comment by slafortune** [comment]
@kavimuru just checking if you had a chance to verify if this is an issue on all platforms.

**Comment by slafortune** [comment]
Thanks, @kavimuru I think that was my misunderstanding.  Are you listing all platforms affected rather than simply checking the ones that are affected?  I remember another GH where there was only one listed (but not checked).

**Comment by kavimuru** [comment]
@slafortune I remove the platforms in which the issue is not reproduced instead of putting check mark.

**Comment by MelvinBot** [comment]
@slafortune this issue was created 2 weeks ago. Are we close to a solution? Let's make sure we're treating this as a top priority. Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
@slafortune Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@slafortune 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by slafortune** [comment]
Sorry for my delay!   I wonder what would be the expected order of listing members - alphabetical always regardless who is the last to reply/send/or be added?

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0128253986dcc547e6

**Comment by MelvinBot** [comment]
Current assignee @slafortune is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @luacmartins (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by abdulrahuman5196** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Group members order changes randomly when it has more than 1 user with phone number.

### What is the root cause of that problem?
First thing to note is the group name is not changing randomly. It will change once you navigate to the phone number user page via the group members list page.
In this flow the these methods will be called in order

1) https://github.com/Expensify/App/blob/4d3390b156fcb59e694b0c21193bd699b276bafd/src/libs/actions/Report.js#L400-L413

2) https://github.com/Expensify/App/blob/4d3390b156fcb59e694b0c21193bd699b276bafd/src/libs/ReportUtils.js#L1375-L1386


Here on the second function if you notice at this specific line

https://github.com/Expensify/App/blob/4d3390b156fcb59e694b0c21193bd699b276bafd/src/libs/ReportUtils.js#L1384

We are calling 'report.participants.sort()' to sort and check if 2 lists are same and return appropriate result. But here as we are directly calling the .sort() function it sorts and mutates the original array resulting in numbers becoming first element as per normal sort logic.

For example: ["test", "test1", "+1", "+2"] will change to ["+1", "+2", "test", "test1"] in the original report.participants object itself. 

As per the current logic we directly display the group members order as we receive from server in the report.participants property. Server returns ["test", "test1", "+1", "+2"] in the above example.


### What changes do you think we should make in order to solve the problem?

We shouldn't use .sort() as it directly mutates the report.participants object. We can use any other sort function which doesn't mutate the object directly and returns new array object for ex: _.sortBy(). There are multiple similar sort functions available we can use any recommended one to solve this here.

https://github.com/Expensify/App/blob/4d3390b156fcb59e694b0c21193bd699b276bafd/src/libs/ReportUtils.js#L1384


### What alternative solutions did you explore? (Optional)

**Comment by sobitneupane** [comment]
Thanks for the proposal @abdulrahuman5196 

The proposal looks good. According to [this SO](https://stackoverflow.com/questions/45126469/native-array-prototype-sort-and-sortby-sorts-values-differently),  `_.sortBy()` and `Array.prototype.sort()` can have different results. So I would suggest using `[...report.participants].sort()`.

cc: @luacmartins

🎀👀🎀 C+ reviewed

**Comment by luacmartins** [comment]
I agree that the proposal is good. I think we should use `_.sortBy`. As mentioned in that SO, the difference comes from identical values, but that shouldn't be the case for participants since they are unique.

**Comment by MelvinBot** [comment]
📣 @abdulrahuman5196 You have been assigned to this job by @luacmartins!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by abdulrahuman5196** [comment]
PR is up - https://github.com/Expensify/App/pull/15509
Should be a quick PR

@sobitneupane  @luacmartins

**Comment by yuwenmemon** [comment]
Heads up @abdulrahuman5196 @luacmartins @sobitneupane @slafortune Seems like our QA team is still seeing this error on staging even with the changes in the PR: https://github.com/Expensify/App/issues/15644

**Comment by abdulrahuman5196** [comment]
@yuwenmemon  could you let us know the reference to the reproducibility? I just tried on staging now it seems to be working for me.

**Comment by abdulrahuman5196** [comment]
@yuwenmemon  I think both the issues are different. Will provide the information on how they are different in a while.

**Comment by abdulrahuman5196** [comment]
@yuwenmemon  @sobitneupane  @luacmartins 
As I suggested before both bugs are different. I am posting the root cause and changes required for the https://github.com/Expensify/App/issues/15644 bug so that we could understand how both the bugs are different. I am posting it here as https://github.com/Expensify/App/issues/15644 is locked for contributors.

You could consider this as proposal as well LOL!!!

## Proposal - https://github.com/Expensify/App/issues/15644

### Please re-state the problem that we are trying to solve in this issue.
To reproduce the bug we don't need to do all the steps mentioned in the bug.
1) Just create a new group chat with phone numbers and names
2) Move to another chat and come back to group chat.
3) The order at step 1 will be changed

### What is the root cause of that problem?

If you notice closely on the video when you create the chat itself the order of the group member was in the order of numbers followed by names. Let me show the code flow and example.

1) User selects list of users in selector page ([+122, a@test.com, +133, b@test.com]) - https://github.com/Expensify/App/blob/main/src/pages/NewChatPage.js#L225
2) It reaches the below code where the participants list is sorted with .sort() (Not sure if it was intentional or not). End order - [+122, +133, a@test.com, b@test.com]
https://github.com/Expensify/App/blob/a7c196fb09e06ed31dfe2f520da134515c835386/src/libs/ReportUtils.js#L1444-L1446
3) Now we pass the above order to openReport API as participants list and same is passed to optimistic report as well - https://github.com/Expensify/App/blob/a7c196fb09e06ed31dfe2f520da134515c835386/src/libs/actions/Report.js#L364
4) It seems if we pass participantsList as emailList param to the OpenReport API for a new group chat creation, it returns the same order in the API response. And the same order is observed for the first time in pusher event.
5) After that when we re-open the group chat we call openReport API to re-fetch new information. Here we don't pass the participantsList and OpenReport API returns the participants order correctly [a@test.com, b@test.com, +122, +133]

Thus, user can see one order until step 4 and another order in step 5 and the order at step 5 is consistently maintained always.

### What changes do you think we should make in order to solve the problem?

1) We should create a sort function which provides result similar to backend sort. I think the backend sort keeps names first and numbers at later part. We can use the new sort function to populate optimistic data on report creation and to OpenReport API for create group chat as well. So the order will be maintained during new group chat creation and openReport API. This way even in offline chat creation it will work the same way.


Additional fixes:
1) Similar to https://github.com/Expensify/App/issues/14886 bug we should remove inplace .sort() and use _.sortBy() so that it won't affect objects outside of the function, at https://github.com/Expensify/App/blob/a7c196fb09e06ed31dfe2f520da134515c835386/src/libs/ReportUtils.js#L1444-L1446

Optional:
1) Backend should provide sorted order for new chat creation in OpenReport API as well.
2) We can also add lint function to avoid .sort() to avoid issues in future

### What alternative solutions did you explore? (Optional)
We can make the backend return participants in the same order of mobile _.sortBy() function. But I am not sure if we want to show numbers first followed by names.

**Comment by abdulrahuman5196** [comment]
As we can see here - https://github.com/Expensify/App/issues/14886#issuecomment-1454822374

https://github.com/Expensify/App/issues/15644 is more to do with group chat creation flow which is not part of this current bug.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.78-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15509


If no regressions arise, payment will be issued on 2023-03-13. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@sobitneupane / @luacmartins] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@sobitneupane / @luacmartins] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@sobitneupane / @luacmartins] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@abdulrahuman5196] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@slafortune] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by abdulrahuman5196** [comment]
@luacmartins  @miljakljajic https://github.com/Expensify/App/issues/15644  is not dupe of this bug as mentioned here - https://github.com/Expensify/App/issues/14886#issuecomment-1454822374

Unable to comment there due to limited access

**Comment by abdulrahuman5196** [comment]
Regression test:

1. Open the app
2. Open any group report with more then 1 user with phone number as primaryLogin 
3. Click on profile Icons of members displayed on top of group report chat
4. Click on one of the phone number user and click on message
5. Navigate back to report archive page
6. Order of the group members shouldn't change.

**Comment by abdulrahuman5196** [comment]
> Job added to Upwork: https://www.upwork.com/jobs/~0128253986dcc547e6

@slafortune  I have applied for the upwork job here.

**Comment by dhanashree-sawant** [comment]
Hi @slafortune, I have also applied to that job for reporting bonus.

**Comment by MelvinBot** [comment]
@slafortune, @luacmartins, @sobitneupane, @abdulrahuman5196 Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by luacmartins** [comment]
Just waiting on payment

**Comment by slafortune** [comment]
regression test GH created and added 👍

**Comment by luacmartins** [comment]
@slafortune are we still waiting for @sobitneupane to accept the payment?

**Comment by sobitneupane** [comment]
I have accepted the offer.

**Comment by luacmartins** [comment]
Cool! Looks like we are all done here!
